### PR TITLE
Add spherical harmonic initial data for CurvedScalarWave

### PR DESCRIPTION
**Pull request by nikwit**
## Proposed changes

Adds the initial data from Scheel(2003), equation 4.1-4.3 for CurvedScalarWave. 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: How about `Mode` of type `std::pair<size_t, int>`?

**Review by nilsvu — COMMENTED** [review]
Will review more later (squash directly if you make the change)

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nikwit) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp]: just saw this, which I will change since `pi` might be moved from...

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: Can you add the ref to Eq. 4.1--4.3 here as well

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: Why delete? No harm in allowing copy, is there?

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: not needed, you can delete this constructor I think.

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: Use `NOLINTNEXTLINE(google-runtime-references)` syntax please

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: Always default-construct classes to an invalid state (NaN in this case). I've had hard-to-find bugs with classes that default-construct to a valid state even when they were forgotten in a pup function, so they did something different than expected.

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: Perhaps note this class is always in 3D?

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp:12]: include Tensor

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp]: include pup

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp]: const radius

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp]: report these errors as `PARSE_ERROR`. They throw, so you can quite easily add a few `CHECK_THROWS_WITH` tests and delete the `LCOV_EXCL_*` directives.

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp]: Just use `x` for the size. While you're at it, just use explicit tensor types instead of the `::type`, it's more readable.

File: tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_PureSphericalHarmonic.cpp
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_PureSphericalHarmonic.cpp]: `<` -> `<=`? (And prefer `++m` just to get in the habit)

File: tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_PureSphericalHarmonic.cpp
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_PureSphericalHarmonic.cpp]: (optional) You can configure the path here so you don't have to write it in `pypp::call`

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nikwit** [comment]
@nilsleiffischer incorporated all your suggestions, please take another look.

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp]: (not requesting this change) Sometimes it can be nice to use structured bindings to name components, like `const auto& [l_mode, m_mode] = mode;`

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: If clang-tidy accepts this, fine. Else delete spaces around dashes.

Delete the `NOLINT` in the line below.

File: tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_PureSphericalHarmonic.cpp
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_PureSphericalHarmonic.cpp]: typo

**Review by nilsvu — CHANGES_REQUESTED** [review]
LGTM please squash directly

**Comment by nikwit** [comment]
squashed

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.hpp]: you haven't deleted the `NOLINT` in the line below

**Review by nilsvu — DISMISSED** [review]

**Comment by kidder** [comment]
this needs a rebase

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/CMakeLists.txt
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/CMakeLists.txt]: remove the extra added lines here

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp:24]: I don't understand why clang-tidy doesn't complain that the `std::move` here is trivial, or about a use-after-move in the next line, or both. Please fix, e.g. use `mode_` in the body of this function.

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp:53]: include `<complex>` I think?

**Review by nilsvu — CHANGES_REQUESTED** [review]
squash directly

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp:24]: Now clang-tidy complains. I think restore the `std::move`. The important thing was to use `mode_` in the body of the function to avoid the use-after-move (still don't understand why clang-tidy didn't complain about that before).

Comment (nikwit) [reply src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp:24]: it appears to be because `std::pair` simply isnt trivially copyable:

https://stackoverflow.com/questions/58283694/why-is-pair-of-const-trivially-copyable-but-pair-is-not

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PureSphericalHarmonic.cpp:24]: The move is fine, but I wonder why clang-tidy didn't complain about the use-after-move.

**Review by nilsvu — APPROVED** [review]

**Comment by nilsvu** [comment]
No need to wait for a second review I think

**Pull request merged**
